chore: update fast bootstrapping (#171)
decentraland/adr: content/ADR-116-fast-bootstrapping.md

# Approach

> Leverage the snapshots mechanism by partitioning the snapshots by time range. The snapshot structure and how it is processed will remain the same. A catalyst will serve the content of all its active entities in multiple snapshots separated by its entity timestamp. Any time a snapshot is processed, a Catalyst remembers this and won't process it again anymore.

Reviewer: ```suggestion
Leverage the snapshots mechanism by partitioning the snapshots by time range. The snapshot structure and how it is processed will remain the same. A Catalyst will serve the content of all its active entities in multiple snapshots separated by its entity timestamp. Any time a snapshot is processed, a Catalyst remembers this and won't process it again anymore.
```

  <img alt="Multiple snapshots" src="resources/RFC-16/multiple_snapshots.png"/>
</figure>

> All the Catalyst will have the same initial timestamp *__t~initial~__* from which they will generate the same time ranges for the snapshots: _[t~initial~, t~1~], [t~1~, t~2~], ...,[t~n-1~, t~n~]_.

Reviewer: ```suggestion
All the Catalysts will have the same initial timestamp *__t~initial~__* from which they will generate the same time ranges for the snapshots: _[t~initial~, t~1~], [t~1~, t~2~], ...,[t~n-1~, t~n~]_.
```

Reviewer: Have a look at how to use subscripts, because it's rendering incorrectly:
![Screenshot 2023-01-11 at 9 38 12 AM](https://user-images.githubusercontent.com/959563/211808207-e1857d1a-57c2-43cf-b26a-758e54044787.png)

Author: I'll review if subscripts can be done or just make them bold

> Given a time range *[t~1~, t~2~]*, a snapshot would contain "_all the entities with entityTimestamp within [t~1~,t~2~] that are **active at the generation time**_". The information associated to an entity in a snapshot are constant data across different Catalysts: _entityId, entityType, scene, pointers, authChain_ and _entityTimestamp_. This is really useful because it implies eventual convergence of the snapshot in different Catalysts. It is eventual because it depends on the _generation time_ of the snapshot.

Reviewer: ```suggestion
Given a time range *[t~1~, t~2~]*, a snapshot would contain "_all the entities with `entityTimestamp` within [t~1~,t~2~] that are **active at the time of generation**_". The information associated to an entity in a snapshot is constant data across all Catalysts: _entityId, entityType, scene, pointers, authChain_ and _entityTimestamp_. This is really useful because it implies eventual convergence of the snapshot in the different Catalysts. It is eventual because it depends on the _generation time_ of the snapshot.
```

### Snapshot process at bootstrap

> A Catalyst asks another one for its snapshots, then it will process each one by deploying all the entities within it, but will save in the database the hash. Next time a Catalyst sees these hashes, it won't process them again.

Reviewer: ```suggestion
A Catalyst asks the other Catalysts in the network for one for the snapshots, then it will process it by deploying all the entities within it, but will save in the database the snapshot hash. Next time a Catalyst sees these hashes, it won't process the same snapshot again.
```

> 1. Reduce the catalyst downtime from ~30 min to less than 5 min.

Reviewer: ```suggestion
1. Reduce the Catalyst downtime from ~30 min to less than 5 min.
```

2. Reduce the network traffic of 1 GB per node to something way smaller.

> 3. Process an active entity only one time.

Reviewer: ```suggestion
3. Process an active entity only once.
```

> 4. Don't generate a big snapshot file making an expensive db query call scanning a huge db table and making high i/o tasks writing the file each time a Catalyst is restarted.

Reviewer: ```suggestion
4. Don't generate a big snapshot file making an expensive db query call that scans the huge db table and avoids making high I/O tasks writing the file each time a Catalyst is restarted.
```

# Key Dependencies and Open Questions

> - How will the transition be where multiple Catalysts have a different behavior in `/snapshots`?

Reviewer: This doesn't look so much like open question (and definitely not a dependency), it seems more like an FAQ kind of question. I would phrase it as a statement (rather than question / answer) and place it at the end of the Approach section.

Author: Changed to FAQ section

>   At the first iteration, the big snasphot will be still generated and served in the `/snapshot` endpoint. If a Catalyst fails to read the snapshots in `/snapshots` it'll make a fallback to the old `/snapshot` endpoint.

Reviewer: ```suggestion
  The first time the Fast Boostrapping code is run, the big snapshot will be still generated and served in the `/snapshot` endpoint. If a Catalyst fails to read the snapshots in `/snapshots` it'll make a fallback to the old `/snapshot` endpoint.
```
